nav2_behavior_tree/include/nav2_behavior_tree/bt_utils.hpp

```diff
@@ -162,6 +169,37 @@ T1 deconflictPortAndParamFrame(
   }
 }
 
+/**
+ * @brief Try reading an import port first, and if that doesn't work
+ * fallback to reading directly the blackboard.
+ * The blackboard must be passed explitly because config() is private in BT.CPP 4.X
+ *
+ * @param bt_node node
+ * @param blackboard the blackboard ovtained with node->config().blackboard
+ * @param param_name std::string
+ * @param behavior_tree_node the node
+ * @return <T>
+ */
+template<typename T> inline
+BT::Result getInputPortOrBlackboard(
+  const BT::TreeNode & bt_node,
+  const BT::Blackboard & blackboard,
+  const std::string & param_name,
+  T & value)
+{
+  if (auto res = bt_node.getInput<T>(param_name, value)) {
+    return res;
+  }
+  if (auto res = blackboard.get<T>(param_name, value)) {
+    return {};
+  }
+  return nonstd::make_unexpected(StrCat(param_name, " not found"));
```

Comment:
There are two different idioms in the C++ world:
1. exceptions and try/catch
2. std::optional and std::expected (soon)

The C++ community is moving steadily toward the second one.
The arguments against exceptions are the runtime overhead and more verbose code to do error handling.
I do use exceptions in BT.CPP but only for... exceptional things! Something that you want to blow up in your face and say "there is an issue in your code, fix it".

An error that will likely happen, should not use exceptions.
This is one of those examples.

Reply:
ok, but I am not adding any try catch in my code, right?

Reply:
what about returning a boolean instead?

Reply:
We run the tree wrapped in a try/except since there are things inside of other BT nodes that can throw (i.e. TF buffers, if control nodes are given incorrect configurations of their childs, etc), so you could throw.

A bool makes sense in this context since either `goal` or `goals` is populated, not both